packages/playground/data-liberation/src/import/WP_Topological_Sorter.php

```diff
@@ -0,0 +1,461 @@
+					id bigint(20) unsigned NOT NULL AUTO_INCREMENT,
+					session_id bigint(20) unsigned NOT NULL,
+					element_type tinyint(1) NOT NULL default %d,
+					element_id text NOT NULL,
+					parent_id text DEFAULT NULL,
+					byte_offset bigint(20) unsigned NOT NULL,
+					sort_order int DEFAULT 1,
+					PRIMARY KEY  (id),
+					KEY session_id (session_id),
+					KEY element_id (element_id(%d)),
+					KEY parent_id (parent_id(%d)),
+					KEY byte_offset (byte_offset)
+				) ' . $wpdb->get_charset_collate(),
+				self::get_table_name(),
+				self::ELEMENT_TYPE_POST,
+				$max_index_length,
+				$max_index_length
+			);
+		}
+
+		require_once ABSPATH . 'wp-admin/includes/upgrade.php';
+		dbDelta( $sql );
+
+		update_option( self::OPTION_NAME, self::DB_VERSION );
+	}
+
+	public static function is_sqlite() {
+		return defined( 'DB_ENGINE' ) && 'sqlite' === DB_ENGINE;
+	}
+
+	/**
+	 * Run in the 'plugins_loaded' action.
+	 */
+	public static function load() {
+		if ( self::DB_VERSION !== (int) get_site_option( self::OPTION_NAME ) ) {
+			// Used to update the database with dbDelta, if needed in the future.
+			self::activate();
+		}
+	}
+
+	/**
+	 * Run by register_deactivation_hook.
+	 */
+	public static function deactivate() {
+		global $wpdb;
+		$table_name = self::get_table_name();
+
+		// Drop the table.
+		$wpdb->query( $wpdb->prepare( 'DROP TABLE IF EXISTS %s', $table_name ) );
+
+		// Delete the option.
+		delete_option( self::OPTION_NAME );
+	}
+
+	/**
+	 * Run by register_uninstall_hook.
+	 */
+	public function reset() {
+		$this->orphan_post_counter = 0;
+		$this->last_post_id        = 0;
+		$this->sorted              = false;
+		$this->current_session     = null;
+		$this->total_categories    = 0;
+		$this->total_posts         = 0;
+		$this->current_item        = 0;
+	}
+
+	/**
+	 * Delete all rows for a given session ID.
+	 *
+	 * @param int $session_id The session ID to delete rows for.
+	 * @return int|false The number of rows deleted, or false on error.
+	 */
+	public function delete_session( $session_id ) {
+		global $wpdb;
+
+		return $wpdb->delete(
+			self::get_table_name(),
+			array( 'session_id' => $session_id ),
+			array( '%d' )
+		);
+	}
+
+	/**
+	 * Map a category to the index.
+	 *
+	 * @param int $byte_offset The byte offset of the category.
+	 * @param array $data The category data.
+	 */
+	public function map_category( $byte_offset, $data ) {
+		global $wpdb;
+
+		if ( empty( $data ) ) {
+			return false;
+		}
+
+		$category_parent = null;
+
+		if ( array_key_exists( 'parent', $data ) && '' !== $data['parent'] ) {
+			$category_parent = $data['parent'];
+		}
+
+		$wpdb->insert(
+			self::get_table_name(),
+			array(
+				'session_id'   => $this->current_session,
+				'element_type' => self::ELEMENT_TYPE_CATEGORY,
+				'element_id'   => (string) $data['term_id'],
+				'parent_id'    => $category_parent,
+				'byte_offset'  => $byte_offset,
+				// Items with a parent has at least a sort order of 2.
```

Comment:
Let's document this more. What's the logic behind the sort_order of 1, 2 etc? Should it be more for deeply nested categories or not?